domain/src/main/java/org/oppia/android/domain/onboarding/DeprecationController.kt

```diff
@@ -0,0 +1,132 @@
+      publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
+    ).invokeOnCompletion { primeFailure ->
+      primeFailure?.let {
+        oppiaLogger.e(
+          "DeprecationController",
+          "Failed to prime cache ahead of data retrieval for DeprecationController.",
+          primeFailure
+        )
+      }
+    }
+  }
+
+  private val deprecationDataProvider by lazy { fetchDeprecationProvider() }
+
+  private fun fetchDeprecationProvider(): DataProvider<DeprecationResponseDatabase> {
+    return deprecationStore.transform(
+      GET_DEPRECATION_RESPONSE_PROVIDER_ID
+    ) { deprecationResponsesDatabase ->
+      DeprecationResponseDatabase.newBuilder().apply {
+        appDeprecationResponse = deprecationResponsesDatabase.appDeprecationResponse
+        osDeprecationResponse = deprecationResponsesDatabase.osDeprecationResponse
+      }.build()
+    }
+  }
+
+  /**
+   * Returns a [DataProvider] containing the the [DeprecationResponseDatabase], which in turn
+   * affects what initial app flow the user is directed to.
+   */
+  fun getDeprecationDatabase(): DataProvider<DeprecationResponseDatabase> = deprecationDataProvider
+
+  /**
+   * Stores a new [DeprecationResponse] to the cache.
+   *
+   * @param deprecationResponse the deprecation response to be stored
+   * @return [AsyncResult] of the deprecation action
+   */
+  fun saveDeprecationResponse(deprecationResponse: DeprecationResponse): DataProvider<Any?> {
+    val deferred = deprecationStore.storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) { deprecationResponseDb ->
+      val deprecationBuilder = deprecationResponseDb.toBuilder().apply {
+        if (deprecationResponse.deprecationNoticeType == DeprecationNoticeType.APP_DEPRECATION)
+          appDeprecationResponse = deprecationResponse
+        else
+          osDeprecationResponse = deprecationResponse
+      }
+        .build()
+      Pair(deprecationBuilder, DeprecationResponseActionStatus.SUCCESS)
+    }
+
+    return dataProviders.createInMemoryDataProviderAsync(ADD_DEPRECATION_RESPONSE_PROVIDER_ID) {
+      return@createInMemoryDataProviderAsync getDeferredResult(deferred)
+    }
+  }
+
+  /**
+   * Retrieves the [DeprecationResponse] from the cache.
+   *
+   * @param deferred a deferred instance of the [DeprecationResponseActionStatus]
+   * @return [AsyncResult]
+   */
+  private suspend fun getDeferredResult(
+    deferred: Deferred<DeprecationResponseActionStatus>
+  ): AsyncResult<Any?> {
+    return when (deferred.await()) {
+      DeprecationResponseActionStatus.SUCCESS -> AsyncResult.Success(null)
+      DeprecationResponseActionStatus.FAILED_TO_STORE_DEPRECATION_RESPONSE -> AsyncResult.Failure(
+        Exception("Failed to store deprecation response")
+      )
+      DeprecationResponseActionStatus.DEPRECATION_RESPONSE_NOT_FOUND -> AsyncResult.Failure(
+        Exception("Deprecation response not found")
+      )
```

Comment:
I noticed that the convention is to define custom exceptions while returning a Deferred Result. PTAL at  [ProfileManagementController](https://github.com/oppia/oppia-android/blob/f7cf96a7b515e1639e97be0ac9311bb0fbc4ccca/domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt#L99) for example

Reply:
Hi Adhiambo. I have been looking at the DeprecationController and I realized that only the `DeprecationResponseActionStatus.SUCCESS `is used. The other two statuses are never used. Can we remove them or do they stay?

Reply:
You can remove them for now. They can be re-added on an as-needed basis in the future.